### ISSUE TITLE
A new random interface

### ISSUE THREAD

**Issue by aarmey** [tensorly/tensorly#436]
Several backends have now implemented the [NumPy random generator interface](https://numpy.org/doc/stable/reference/random/generator.html). With this, we can rethink how random state is passed throughout the random module.

It may also be helpful to think about how context could be passed in as a property of the random generator. For example, right now the only option to extend methods like `random_cp()` and `random_tucker()` to cover non-negative or complex cases would be to add redundant options to each method. However, these properties really just concern themselves with how the random numbers are generated. If these could be wrapped into the number generator process, then other possibilities, such as changing how the numbers are distributed, would be possible through a uniform interface. The `random_*` methods would then only concern themselves with the tensor structural form.

**Comment by JeanKossaifi** [comment]
Agreed, though generating random factorized tensors is involved if we want to maintain some properties on the reconstructed tensor (see e.g. in tensorly-torch, we allow to generate random tensors such that the rec. tensor has `0 mean` and a given `std`.

**Comment by aarmey** [comment]
At least how I envision it, both `random_*()` and the random number generating object would be able to modify the result. For instance, if you want to change the random factors to be Gamma or Beta distributed, you would set that in the number generating object. However, the `random_*()` methods would still handle factor-wide modifications such as orthogonality. From looking through tensorly-torch, I think it looks like you set the mean and stdev by z-scoring the factors? If so, you could still do that at the `random_*()` after the number generation has occurred.
